### DISCUSSION TITLE
Why are there different formats for dirtyFields?

### DISCUSSION THREAD

**Discussion by Kikkomann** [react-hook-form/react-hook-form#8818] [Q&A]
I get two different `formstate.dirtyFields` obejct types based on two different situations. Let me try to explain:

I have a `FieldArray` that consists of custom components:

```
[
  { material: Wood, type: new, amount: 10 }
  { material: Plastic, type: reused, amount: 1 }
]
```

If I change material in the second item, my `formState.dirtyFields` will look like this:
`[ empty/null, { material: true } ]`. All good, then I can just send an update on all the indices that exist.

The thing is that when I add a new item to the `FieldArray`, my `dirtyFields` now look like this:

```
[
  { material: false, type: false, amount: false }
  { material: true, type: false, amount: false }
  { material: true, type: true, amount: true }
]
```

The dirty fields are correct, but this is another format, which makes it cumbersome for me to figure out which items I need to update.

Is it a bug, am I doing something wrong or is there a meaning of having two different formats that I just don't see?

Reproducible sandbox:
https://codesandbox.io/s/dirty-fields-problem-lqk3dy

First edit a field and see the printed `dirtyFields`.
Then add an item to see that the printed fields look different.

**Comment by bluebill1049** [comment]
It's not a bug, it's the most efficient way to mark dirty fields for field array to include both dirty and none dirty, unless someone can come up with a better algorithm to improve the runtime and omit those `false` fields.

**Comment by Kikkomann** [comment]
It is not the format of a `dirtyField` that is the issue. It is that I get two different formats on the `dirtyFields`. I guess it is only a problem when using `FieldArray`s, so one is where I get `empty` for the items in the array that has not been changed and only the changed fields for all the items that have changed fields. The other format is, as you say, all items in `dirtyFields` with true/false on each field in each item.

```
// Format one:
[
  empty/null,
  { material: true }
]

// Format two:
[
  { material: false, type: false, amount: false }
  { material: true, type: false, amount: false }
  { material: true, type: true, amount: true }
]
```

The two formats above, I get for the same list of items. First I only update the `material` field and get format one, then I append a new item in the `FieldArray` and get format two.
